internal/api/signup.go

```diff
@@ -161,15 +161,26 @@ func (a *API) Signup(w http.ResponseWriter, r *http.Request) error {
 		user, err = models.IsDuplicatedEmail(db, params.Email, params.Aud, nil)
 	case "phone":
 		if !config.External.Phone.Enabled {
-			return badRequestError("Phone signups are disabled")
+			return badRequestError(ErrorCodePhoneProviderDisabled, "Phone signups are disabled")
 		}
 		params.Phone, err = validatePhone(params.Phone)
 		if err != nil {
 			return err
 		}
 		user, err = models.FindUserByPhoneAndAudience(db, params.Phone, params.Aud)
 	default:
-		return invalidSignupError(config)
+		msg := ""
+		if config.External.Email.Enabled && config.External.Phone.Enabled {
+			msg = "Sign up only available with email or phone provider"
+		} else if config.External.Email.Enabled {
+			msg = "Sign up only available with email provider"
+		} else if config.External.Phone.Enabled {
+			msg = "Sign up only available with phone provider"
+		} else {
+			msg = "Sign up with this provider not possible"
+		}
+
+		return badRequestError(ErrorCodeValidationFailed, msg)
```

Comment:
This should probably be an unprocessable entity error as it's not validation on the body, but a state of the system.

Reply:
Actually no. The reason why it's a 400 validation_failed is because:

1. The payload `params.Provider` does not contain a recognized value.
2. The error message is a developer aid, it's telling the developer that they _can_ use `phone` or `email` (depending on configuration); but a production application should never run into this issue as it means the developer has miconfigured their server and an end-user can't do anything about it.